timemachine/cpp/src/context.cu

```diff
@@ -50,6 +64,342 @@ Context::~Context() {
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else {
+        throw std::runtime_error("Unable to cast potential to NonbondedAllPairs");
+    }
+}
+
+bool is_nonbonded_potential(std::shared_ptr<Potential> pot) {
+    if (std::shared_ptr<NonbondedAllPairs<float, true>> nb_pot =
+            std::dynamic_pointer_cast<NonbondedAllPairs<float, true>>(pot);
+        nb_pot) {
+        return true;
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        return true;
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        return true;
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        return true;
+    }
+    return false;
+}
+
+void set_nonbonded_potential_idxs(
+    std::shared_ptr<Potential> pot, const int num_idxs, const unsigned int *d_idxs, const cudaStream_t stream) {
+    if (std::shared_ptr<NonbondedAllPairs<float, true>> nb_pot =
+            std::dynamic_pointer_cast<NonbondedAllPairs<float, true>>(pot);
+        nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else {
+        throw std::runtime_error("unable to cast potential to NonbondedAllPairs");
+    }
+}
+
+void Context::_assert_temperature_matches(const double temperature) {
```

Comment:
nit: name sounds general, but this is specific to local MD?

Reply:
Right now it is specific to local MD. But in any case where MD moves are dependent on temperature, we will want to verify that the thermostat agrees. Could rename in the mean time.